Fix memory leak when using streams in a microtask loop in Node.js

twiss opened the pull request:
Hi @MattiasBuelens! First of all, thanks for all your work on this polyfill.

In Node.js, when using streams in a microtask loop that never finishes (e.g. `while (true) { await ... }`, or on a server that's constantly receiving events) there's a memory leak.

AFAIU, it comes from an internal list of unhandled promise rejections in Node, which retains promises created by this library, which in turn retains the closure in which they were created, which in turn retains the data that the stream was created from in some cases.
https://github.com/nodejs/node/blob/d18b0a01320e392e48f0c475effd8d4db2bb71a8/lib/internal/process/promises.js#L62

Node cleans up this list at the end of every macrotask loop tick, but if there are always new microtasks coming in it never gets to that.
https://github.com/nodejs/node/blob/d18b0a01320e392e48f0c475effd8d4db2bb71a8/lib/internal/process/next_tick.js#L82
(I believe this is more or less as per spec, as unhandled rejections are supposed to be thrown as a macrotask, not a microtask. There's probably nothing preventing Node from cleaning up _handled_ rejections in the microtask loop, though, so it could be argued that this is a Node bug.)

As `promiseRejectHandler` is called synchronously, even `Promise.reject(reason).catch(() => {})` adds to this list.
This PR essentially replaces that by:
```js
let reject;
let promise = new Promise((res, rej) => {
  reject = rej;
});
promise.catch(() => {});
reject(reason);
return promise;
```

which makes sure that `promiseRejectHandler` is never called, by catching the promise before it's rejected.

---

Chrome doesn't seem to leak memory, neither with its built-in streams (which use the internal `v8.markPromiseAsHandled` function) nor with this polyfill, so maybe this should be fixed in Node as well, but I'd rather fix it here for now and not have to wait for a new Node release :)

Firefox seems to also have a memory leak that is similar to this but not exactly the same - even with this patch, and even when running a _macrotask_ loop (e.g. with `setTimeout`), it still leaks memory. A memory snapshot points to the same sort of retention chain, with the addition that each Promise points to the previous one (PromiseReactionRecord -> Promise -> (...) -> PromiseReactionRecord -> Promise -> Function -> LexicalEnvironment).

MattiasBuelens commented:
Interesting find!

If I understand correctly: if a promise is rejected **before** a rejection handler is attached, Node has to queue a task to check if the rejection is truly unhandled. But if we make sure to attach a rejection handler **before** we reject the promise, then Node knows that it will definitely be handled and can skip queuing such a task?

twiss commented:
Hmhm, exactly.

twiss commented:
Thanks!

> Found one place where a `catch` handler was still attached where it is no longer needed, so I removed that.

:+1:

> I also restored the verbose logs in `writable-stream`. I know that we don't actually use those for anything, and I should probably remove them _at some point_, but I like to keep this PR focused on just the memory leak. 🙂

Alrighty. The only reason I removed them was that since the functions call each other now, you would end up with logs like

```
defaultWriterReadyPromiseInitializeAsResolved()
defaultWriterReadyPromiseInitialize()
defaultWriterReadyPromiseResolve()
```

which might be a bit confusing, but probably not a big deal either.

Thanks for the quick response & merge!

MattiasBuelens commented:
> Alrighty. The only reason I removed them was that since the functions call each other now, you would end up with logs like
> 
> ```
> defaultWriterReadyPromiseInitializeAsResolved()
> defaultWriterReadyPromiseInitialize()
> defaultWriterReadyPromiseResolve()
> ```
> which might be a bit confusing, but probably not a big deal either.

Ah, I see, that makes sense. I should just remove *all* those verbose logs at some point... 😛 

I'll probably release this as part of `0.3.0` later today, together with the TypeScript rewrite in #7. 😁
